src/main/java/org/folio/rest/impl/InventoryHelper.java

```diff
@@ -78,25 +101,68 @@
       .thenApply(compPOL::withInstanceId);
   }
 
-  public CompletableFuture<String> getOrCreateHoldingsRecord(CompositePoLine compPOL) {
+  /**
+   * Returns list of item id's corresponding to given PO line.
+   * Items are either retrieved from Inventory or new ones are created if no corresponding item records exist yet.
+   *
+   * @param compPOL   PO line to retrieve/create Item Records for. At this step PO Line must contain instance Id
+   * @return future with list of item id's
+   */
+  public CompletableFuture<List<String>> handleItemRecords(CompositePoLine compPOL) {
+    List<CompletableFuture<List<String>>> itemsPerHolding = new ArrayList<>();
+
+    // Group all locations by location id because the holding should be unique for different locations
+    groupLocationsById(compPOL)
+      .forEach((locationId, locations) -> {
+        int expectedQuantity = calculateInventoryItemsQuantity(compPOL, locations);
+        // For some cases items might not be created e.g. Electronic resource with create inventory set to false
+        if (expectedQuantity > 0) {
+          itemsPerHolding.add(
+            // Search for or create a new holding and then create items for this holding
+            getOrCreateHoldingsRecord(compPOL, locationId)
+              .thenCompose(holdingId -> handleItemRecords(compPOL, holdingId, expectedQuantity))
+          );
+        }
+      });
+
+    return collectResultsOnSuccess(itemsPerHolding);
+  }
+
+  /**
+   * Wait for all requests completion and collect all resulting objects In case any failed, complete resulting future with the exception
+   * @param futures list of futures and each produces list of resulting objects on completion
+   * @param <T> resulting type
+   * @return resulting objects
+   */
+  private <T> CompletableFuture<List<T>> collectResultsOnSuccess(List<CompletableFuture<List<T>>> futures) {
```

Comment:
In general I like the idea of pulling out and generalizing this code as it's a common pattern used throughout the class (and module).  Unfortunately, it seems that minor variations prevent this from being used in at least one other case in this class.  I wonder if that indicates that you've stopped just short of making this truly useful, or if we should consider moving this to somewhere in the code which allows it to be easily called from other parts of the code.

See lines 399 - 404

Reply:
Just food for thought... this isn't something that should prevent this from being merged.

Reply:
I moved `collectResultsOnSuccess` to `HelperUtils` and updated other classes to use it.